File: src/tools/evaluate_bias.ts
// src/tools/evaluate_bias.ts
import { Tool } from '@modelcontextprotocol/sdk/types.js';
import { PythonBridge } from '../python_bridge.js';

export const evaluateBiasTool: Tool = {
  name: 'evaluate_bias',
  description: 'Evaluates text, code, or data for bias against protected attributes. Use content_type="code" for source code analysis (comments, variable names, algorithmic bias). Use content_type="text" for natural language content. Supports single attribute (backward compatible) or multiple attributes via protected_attributes array.',
  inputSchema: {
    type: 'object',
    properties: {
      content: {
        type: 'string',
        description: 'The text, code, or dataset row to evaluate for bias',
      },
      protected_attribute: {
        type: 'string',
        description: 'The protected attribute to check (e.g., "gender", "race", "age"). Use this for single attribute (backward compatible).',
        enum: ['gender', 'race', 'age', 'disability'],
      },
      protected_attributes: {
        type: 'array',
        items: { type: 'string', enum: ['gender', 'race', 'age', 'disability'] },
        description: 'Array of protected attributes to check simultaneously. If provided, will check multiple attributes and detect intersectional bias.',
      },
      task_type: {
        type: 'string',
        description: 'The type of task being evaluated',
        enum: ['generative', 'classification'],
      },
      content_type: {
        type: 'string',
        description: 'Type of content: "text" for natural language, "code" for source code',
        enum: ['text', 'code'],
        default: 'text',
      },
    },
    required: ['content', 'task_type'],
  },
};

export async function handleEvaluateBias(
  bridge: PythonBridge,
  args: any
): Promise<any> {
  const protectedAttributes = args.protected_attributes as string[] | undefined;
  const protectedAttribute = args.protected_attribute as string | undefined;
  
  let result;
  if (protectedAttributes && Array.isArray(protectedAttributes)) {
    result = await bridge.evaluateBias(
      args.content as string,
      protectedAttribute || '',
      args.task_type as 'generative' | 'classification',
      (args.content_type as 'text' | 'code') || 'text',
      protectedAttributes
    );
  } else {
    result = await bridge.evaluateBias(
      args.content as string,
      protectedAttribute as string,
      args.task_type as 'generative' | 'classification',
      (args.content_type as 'text' | 'code') || 'text'
    );
  }
  
  return {
    content: [
      {
        type: 'text',
        text: JSON.stringify(result, null, 2),
      },
    ],
  };
}

File: py_engine/tools/evaluate_bias.py
# py_engine/tools/evaluate_bias.py
"""
Tool handler for evaluate_bias MCP tool.
Evaluates text, code, or data for bias against protected attributes.
"""
from typing import Dict, Any
from core.auditor import (
    evaluate_bias_audit,
    evaluate_multi_attribute_bias,
)
from core.code_auditor import evaluate_code_bias
from models import EvaluateBiasRequest


def handle_evaluate_bias(req: EvaluateBiasRequest) -> Dict[str, Any]:
    """
    Handle evaluate_bias tool request.
    
    Args:
        req: Validated EvaluateBiasRequest
        
    Returns:
        Response dictionary with result
    """
    protected_attributes = req.protected_attributes
    protected_attribute = req.protected_attribute
    
    # Support both single attribute (backward compatible) and multiple attributes
    if protected_attributes is not None and len(protected_attributes) > 0:
        # Multiple attributes - use multi-attribute evaluation if > 1
        if len(protected_attributes) > 1:
            result = evaluate_multi_attribute_bias(
                req.content, 
                protected_attributes, 
                req.task_type, 
                req.content_type
            )
            return {'result': result}
        else:
            # Single attribute in list format
            target_attr = protected_attributes[0]
    else:
        target_attr = protected_attribute or ''
    
    # Single attribute evaluation
    if req.content_type == 'code':
        result = evaluate_code_bias(req.content, target_attr)
    else:
        result = evaluate_bias_audit(req.content, target_attr, req.task_type)
    
    return {'result': result}

